src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,117 @@ public SubstitutionMatrix(final byte[] matrix) {
+     * @param refBase reference base being substituted
+     * @param code substitution code
+     * @return base to be substituted for this (refBase, code) pair
+     */
+    public byte base(final byte refBase, final byte code) {
+        final byte base = baseByCode[refBase][code];
+        if (base == NO_BASE) {
+            // attempt to retrieve a code for a reference base that isn't in the substitution matrix
+            throw new IllegalArgumentException(String.format("CRAM: Attempt to retrieve a substitution base for '%c'", (char) refBase));
+        }
+        return base;
+    }
+
     /**
      * Return this substitution matrix as a byte array in a form suitable for serialization.
-     * @return
+     * @return the encoded matrix in serialized form
      */
     public byte[] getEncodedMatrix() {
         return encodedMatrixBytes;
     }
 
+    @Override
+    public String toString() {
+        final StringBuilder stringBuilder = new StringBuilder();
+        for (final SubstitutionBase r : SubstitutionBase.values()) {
+            stringBuilder.append((char) r.getBase());
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[r.getBase()][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        // lower case substitutions
+        for (final SubstitutionBase r : SubstitutionBase.values()) {
+            stringBuilder.append(Character.toLowerCase((char)r.getBase()));
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[Character.toLowerCase((char)r.getBase())][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        return stringBuilder.toString();
+    }
+
+    // Initialize the baseByCode array to a sentinel value
+    private void initializeBaseByCode() {
+        for (final byte[] base : baseByCode) {
+            Arrays.fill(base, NO_BASE);
+        }
+    }
+
+    // populate a matrix of substitution frequencies from a list of CramCompressionRecords with Substitution features
+    private static long[][] buildFrequencies(final List<CramCompressionRecord> cramRecords) {
+        final long[][] frequencies = new long[SYMBOL_SPACE_SIZE][SYMBOL_SPACE_SIZE];
+        for (final CramCompressionRecord record : cramRecords) {
+            if (record.readFeatures != null) {
+                for (final ReadFeature readFeature : record.readFeatures) {
+                    if (readFeature.getOperator() == Substitution.operator) {
+                        final Substitution substitution = ((Substitution) readFeature);
+                        final byte refBase = substitution.getReferenceBase();
+                        final byte base = substitution.getBase();
+                        frequencies[refBase][base]++;
```

Comment:
refBase and base are typed as bytes... they won't ever take values over 127 (as the highest byte is set to 1 then they are negative). However frequencies is a [256][256] 2d array. ... perhaps despite than there are 256 bytes you actually wan to create this arrays to only 128 in size and simply disregard negative base values? Need to check in other places but is quite possible that this is a common problem with other arrays around here like codeByByte and byteByCode.

Reply:
True. I reduced the size to 128 and added guards to check for negative values which we shuld never see.

Reply:
Also, in the next round I'll add similar (non-negative) restrictions to the `Substitution` code.